src/deepLink/deepLinkService.ts

```diff
@@ -0,0 +1,403 @@
+			[DeepLinkServiceActions.RepoMatchedWithId]: DeepLinkServiceStates.RemoteMatch,
+			[DeepLinkServiceActions.RepoMatchedWithRemoteUrl]: DeepLinkServiceStates.TargetMatch,
+			[DeepLinkServiceActions.DeepLinkErrored]: DeepLinkServiceStates.Idle,
+		},
+		[DeepLinkServiceStates.RemoteMatch]: {
+			[DeepLinkServiceActions.RemoteMatched]: DeepLinkServiceStates.TargetMatch,
+			[DeepLinkServiceActions.RemoteMatchFailed]: DeepLinkServiceStates.AddRemote,
+		},
+		[DeepLinkServiceStates.AddRemote]: {
+			[DeepLinkServiceActions.RemoteAdded]: DeepLinkServiceStates.OpenGraph,
+			[DeepLinkServiceActions.DeepLinkErrored]: DeepLinkServiceStates.Idle,
+			[DeepLinkServiceActions.DeepLinkCanceled]: DeepLinkServiceStates.Idle,
+		},
+		[DeepLinkServiceStates.TargetMatch]: {
+			[DeepLinkServiceActions.TargetIsRemote]: DeepLinkServiceStates.OpenGraph,
+			[DeepLinkServiceActions.TargetMatched]: DeepLinkServiceStates.OpenGraph,
+			[DeepLinkServiceActions.TargetMatchFailed]: DeepLinkServiceStates.Fetch,
+		},
+		[DeepLinkServiceStates.Fetch]: {
+			[DeepLinkServiceActions.TargetFetched]: DeepLinkServiceStates.FetchedTargetMatch,
+			[DeepLinkServiceActions.DeepLinkErrored]: DeepLinkServiceStates.Idle,
+			[DeepLinkServiceActions.DeepLinkCanceled]: DeepLinkServiceStates.Idle,
+		},
+		[DeepLinkServiceStates.FetchedTargetMatch]: {
+			[DeepLinkServiceActions.TargetMatched]: DeepLinkServiceStates.OpenGraph,
+			[DeepLinkServiceActions.DeepLinkErrored]: DeepLinkServiceStates.Idle,
+		},
+		[DeepLinkServiceStates.OpenGraph]: {
+			[DeepLinkServiceActions.DeepLinkResolved]: DeepLinkServiceStates.Idle,
+			[DeepLinkServiceActions.DeepLinkErrored]: DeepLinkServiceStates.Idle,
+		},
+	};
+
+	constructor(private readonly container: Container) {
+		this._state = DeepLinkServiceStates.Idle;
+		this._disposables = [
+			container.uri.onUri((event: UriEvent) => {
+				if (event.type === UriTypes.DeepLink && this._state === DeepLinkServiceStates.Idle) {
+					if (!event.repoId || !event.linkType || !event.uri || !event.remoteUrl) {
+						void window.showErrorMessage(`Error resolving deep link: missing required properties.`);
+						return;
+					}
+
+					if (!Object.values(DeepLinkTypes).includes(event.linkType)) {
+						void window.showErrorMessage(`Error resolving deep link: unknown link type.`);
+						return;
+					}
+
+					if (event.linkType !== DeepLinkTypes.Remote && !event.targetId) {
+						void window.showErrorMessage(
+							`Error resolving deep link of type ${event.linkType}: no target id provided.`,
+						);
+						return;
+					}
+
+					this._repoId = event.repoId;
+					this._targetType = event.linkType;
+					this._uri = event.uri;
+					this._remoteUrl = event.remoteUrl;
+					this._targetId = event.targetId;
+
+					this._stateChange.fire({
+						state: this._state,
+						action: DeepLinkServiceActions.DeepLinkEventFired,
+					});
+				}
+			}),
+			this._stateChange.event(async (serviceStateChange: DeepLinkServiceStateChange) => {
```

Comment:
I'm a bit worried about this vs a `while` loop, because when VS Code `EventEmitter`s "fire" they aren't async "events" they are more pure function calls, so this could cause a very deep stack nesting for a whole set of state transitions (since every transition will still be on the call stack)

Reply:
Alternatively, without changing much we could use `queueMicroTask` before calling "fire" which would allow the stack to unwind

Reply:
Updated in commit: [Use while loop for deep link processing](https://github.com/gitkraken/vscode-gitlens/pull/2471/commits/de4588cb00e5878bfcfad8d7b14f0ce2138dd724)